packages/cli/src/chat/chat-server.ts

```diff
@@ -1,27 +1,70 @@
 import { Service } from '@n8n/di';
+import { OnShutdown } from '@n8n/decorators';
 import type { Application } from 'express';
 import type { Server as HttpServer } from 'http';
 import { ServerResponse } from 'http';
 import { parse as parseUrl } from 'url';
 import type { WebSocket } from 'ws';
 import { Server as WebSocketServer } from 'ws';
 
+import { parse as parseQuery } from 'querystring';
+
 import { ChatService } from './chat-service';
 import type { ChatRequest } from './chat-service.types';
 
 interface ExpressApplication extends Application {
 	handle: (req: any, res: any) => void;
 }
 
+type UpgradeRecord = { count: number; lastAttempt: number };
+
+const UPGRADE_LIMIT = 20;
+const UPGRADE_WINDOW = 60 * 1000;
+
 @Service()
 export class ChatServer {
 	private readonly wsServer = new WebSocketServer({ noServer: true });
 
+	private readonly upgradeAttempts = new Map<string, UpgradeRecord>();
+
 	constructor(private readonly chatService: ChatService) {}
 
 	setup(server: HttpServer, app: Application) {
 		server.on('upgrade', (req: ChatRequest, socket, head) => {
-			if (parseUrl(req.url).pathname?.startsWith('/chat')) {
+			const parsedUrl = parseUrl(req.url ?? '');
+			const params = parseQuery(parsedUrl.query ?? '');
+
+			const { sessionId, executionId } = params;
+
+			if (!sessionId || !executionId) {
+				return;
+			}
```

Comment:
Sorry I actually got a bit mixed how we do things in the task runner's broker. Also didn't realize that push connection uses the same server. Now seeing this the old implementation was actually better (where the params were checked in `chat-service`). Can we go back to taht 🙈